### PR TITLE
8336635: Add IR test for Reference.refersTo intrinsic

### PR DESCRIPTION
**Pull request by shipilev**
As the follow-up for [JDK-8256999](https://bugs.openjdk.org/browse/JDK-8256999), we can directly test for the (absence of) GC barriers in C2 `Reference.refersTo` intrinsic. This would be come less relevant for G1 after Late Barrier expansion lands, but would still be relevant for Shenandoah for a while. It would also be good for backports.

Additional testing:
 - [x] Test fails when I deliberately remove `AS_NO_KEEPALIVE` from the C2 intrinsic
 - [x] Test passes with GC explicitly passed as `TEST_VM_OPTS`

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8336635](https://bugs.openjdk.org/browse/JDK-8336635): Add IR test for Reference.refersTo intrinsic (**Enhancement** - P4)

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) 🔄 Re-review required (review applies to [ca1ad60e](https://git.openjdk.org/jdk/pull/20215/files/ca1ad60e5c3de3e1a2ae82aff624fe28a0332b39))

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20215/head:pull/20215` \
`$ git checkout pull/20215`

Update a local copy of the PR: \
`$ git checkout pull/20215` \
`$ git pull https://git.openjdk.org/jdk.git pull/20215/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20215`

View PR using the GUI difftool: \
`$ git pr show -t 20215`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20215.diff">https://git.openjdk.org/jdk/pull/20215.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20215#issuecomment-2232915332)

### CODE REVIEW

File: test/hotspot/jtreg/compiler/c2/irTests/gc/ReferenceRefersToTests.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+import jdk.test.lib.Asserts;
+import compiler.lib.ir_framework.*;
+import jdk.test.whitebox.gc.GC;
+
+import java.lang.ref.*;
+
+/*
+ * @test
+ * @bug 8336635
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/irTests/gc/ReferenceRefersToTests.java:33]: Should this refer to `8256999` instead?

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise. I submitted this through our testing and will report back once it passed.

File: test/hotspot/jtreg/compiler/c2/irTests/gc/ReferenceRefersToTests.java (hunk 1 above)
Comment (shipilev) [reply test/hotspot/jtreg/compiler/c2/irTests/gc/ReferenceRefersToTests.java:33]: Right. It makes a bit more sense to reference the issue that added the C2 intrinsics.

Comment (shipilev) [reply test/hotspot/jtreg/compiler/c2/irTests/gc/ReferenceRefersToTests.java]: Fixed in new commit.

**Review by TobiHartmann — APPROVED** [review]

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by TobiHartmann** [comment]
The new test failed with `-ea -esa -XX:CompileThreshold=100 -XX:+UnlockExperimentalVMOptions -server -XX:-TieredCompilation`

```
Failed IR Rules (3) of Methods (3)
----------------------------------
1) Method "public boolean compiler.c2.irTests.gc.ReferenceRefersToTests.phantom_null()" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#LOAD#_", "1"}, applyIfPlatformOr={}, failOn={}, applyIfPlatform={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(Load(B|UB|S|US|I|L|F|D|P|N).*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

2) Method "public boolean compiler.c2.irTests.gc.ReferenceRefersToTests.soft_null()" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#LOAD#_", "1"}, applyIfPlatformOr={}, failOn={}, applyIfPlatform={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(Load(B|UB|S|US|I|L|F|D|P|N).*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

3) Method "public boolean compiler.c2.irTests.gc.ReferenceRefersToTests.soft_ref()" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#LOAD#_", "1"}, applyIfPlatformOr={}, failOn={}, applyIfPlatform={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(Load(B|UB|S|US|I|L|F|D|P|N).*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
```

Here are the logs:
[fail1.txt](https://github.com/user-attachments/files/16302499/fail1.txt)
[fail2.txt](https://github.com/user-attachments/files/16302504/fail2.txt)

**Comment by shipilev** [comment]
Thanks! Looking...

**Comment by shipilev** [comment]
Reproduced the failure locally. The failure happens when C2 does not inline the code aggressively, so we do not reach the intrinsics. Amended the test to make it more reliable. Does not fail with 1000 repetitions. Please test again, @TobiHartmann? Thanks.

**Comment by shipilev** [comment]
@TobiHartmann, could you please run this version through your tests?

**Comment by TobiHartmann** [comment]
Sorry, this slipped through! @chhagedorn re-submitted testing. We will report back once it passed.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by shipilev** [comment]
> All tests passed.

Thanks! Please approve?

**Comment by shipilev** [comment]
Thanks!

/integrate

**Pull request closed without merging**